### PR TITLE
[ENH] Add Hidalgo annotation

### PR DESCRIPTION
**Pull request by KatieBuc**
#### Reference Issues/PRs
Drafting a template for the Hidalgo algorithm, an estimator in the annotation module (issue #2813). Ideas will be discussed as a community about an annotation base class object, what it could look like.

#### What does this implement/fix? Explain your changes.
Introduces a new concept/algorithm for segmentation given a (multivariate) timeseries.

intrinsdim.py -> output is scalar value representing intrinsic dimensionality
hidalgo.py -> extends the above model, assuming the data exists on a mixture of manifolds (each with corresponding dimension), output is a vector of size `N x 1` corresponding to each the multivariate input of size `N x c`. For example, if we assume there are `K=3` manifolds then an output vector for `N=10` would look like `(d_1, d_3, d_2, d_2, d_2, d_3, d_1, d_1, d_1, d_1)`, where `d_k` is the dimensionality for manifold `K`.

#### Does your contribution introduce a new dependency? If yes, which one?
TDB.
Hidalgo relies on Gibbs sampling (MCMC) with a custom likelihood function in the model. Specifically, it is assumed that the neighborhood of a point is more likely to contain points from the same manifold than from different manifolds. This requirement is enforced with an additional term in the likelihood that penalizes local inhomogeneity [points belonging to different manifolds]. 

Options for implementing the Gibbs sampler include:
- including the C code from https://github.com/micheleallegra/Hidalgo/blob/master/python/gibbs.c
- re-writing the model and Gibbs sampler in python
- interfacing pymc, seeing if our model can be fed into an 'out-of-the-box' Gibbs method

### Notes
References: https://arxiv.org/abs/1902.10459

- [x] Gibbs sampler code from [this repository](https://github.com/micheleallegra/Hidalgo) has been rewritten (c++ to python)
- [x] Unit tests written as reference
- [x] Paired programming
- [x] Example notebook created

@lmmentel @Lovkush-A @fkiraly @miraep8

### CODE REVIEW

**Comment by aiwalter** [comment]
@KatieBuc is this ready for review? :)

**Comment by KatieBuc** [comment]
@aiwalter why yes, yes it is*

*currently running all tests, assuming all is well or that I'm busy fixing things in the next few hours/day.

File: sktime/annotation/hidalgo.py [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+            Ground truth annotations for training, not specified as HidAlgo
+            is unsupervised.
+
+        Returns
+        -------
+        self
+        """
+        seed = self.seed
+        self._rng = check_random_state(seed)
+        return self
```
Comment (aiwalter) [inline sktime/annotation/hidalgo.py:583]: lets remove `_fit` then and add the tag `"fit_is_empty": True` and add a line `self._rng = check_random_state(self.seed)` there

File: sktime/annotation/hidalgo.py [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+            * (1 - zeta) ** (q - q1)
+            for q1 in range(q + 1)
+        ]
+    )
+
+
+class Hidalgo(BaseSeriesAnnotator):
+    """Class to fit parameters of the HidAlgo intrinsic dimension model.
+
+    explain, reference
```
Comment (aiwalter) [inline sktime/annotation/hidalgo.py:60]: yes would be nice to have a simple description of that this algo is actually doing, so that people like me can also understand it :)

**Review by aiwalter — COMMENTED** [review]
@KatieBuc great work, left some comments

File: sktime/annotation/hidalgo.py [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+__all__ = ["Hidalgo"]
+
+
+from functools import reduce
+
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+from sklearn.utils.validation import check_random_state
+
+
```
[the remark this answers is not in the document]
Comment (KatieBuc) [reply sktime/annotation/hidalgo.py:34]: did you mean to rearrange the imports? I've changed it to
```
from sktime.annotation.base import BaseSeriesAnnotator
from sklearn.neighbors import NearestNeighbors
from sklearn.utils.validation import check_random_state
from functools import reduce
import numpy as np
```

Let me know if I misunderstood

File: sktime/annotation/hidalgo.py [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+    Examples
+    --------
+    >>> from sktime.annotation.hidalgo import Hidalgo
+    >>> import numpy as np
+    >>> np.random.seed(123)
+    >>> X = np.random.rand(10,3)
+    >>> X[6:, 1:] = 0
+    >>> model = Hidalgo(K=2, Niter=50, seed=10)
+    >>> fitted_model = model._fit(X)
+    >>> Z = fitted_model._predict(X)
```
[the remark this answers is not in the document]
Comment (KatieBuc) [reply sktime/annotation/hidalgo.py:121]: Yeah, I agree and was confused by this. In BaseSeriesAnnotator:
```
    def fit(self, X, Y=None):
        """Fit to training data.

        Parameters
        ----------
        X : pd.DataFrame
            Training data to fit model to (time series).
        Y : pd.Series, optional
            Ground truth annotations for training if annotator is supervised.

        Returns
        -------
        self :
            Reference to self.

        Notes
        -----
        Creates fitted model that updates attributes ending in "_". Sets
        _is_fitted flag to True.
        """
```

So I changed it from `y` to `Y` to avoid a unit test failure I was getting. It's not calling in any `y` now, just as a temporary fix until we redesign this BaseClass

File: sktime/annotation/hidalgo.py [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+        f1[0] = f[0] + N_in
+        f1[1] = f[1] + N * q - N_in
+
+        return N_in, f1
+
+    def get_random_z(self):
+        """Generate random Z from random number generator."""
+        K = self.K
+        N = self.N
+        return self._rng.randint(0, K, N)
```
[the remark this answers is not in the document]
Comment (KatieBuc) [reply sktime/annotation/hidalgo.py:259]: good point - we do not!

File: sktime/annotation/hidalgo.py (hunk 1 above)
Comment (KatieBuc) [reply sktime/annotation/hidalgo.py]: wait, put `self._rng = check_random_state(self.seed)` where?

File: sktime/annotation/hidalgo.py [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+        Iout_track = np.append(0, Iout_track[:-1]).astype(int)
+        Iin = Iin[:, 1:]
+        Iin = np.reshape(Iin, (N * q,)).astype(int)
+
+        self.N = N
+        self.mu = mu
+        self.Iin = Iin
+        self.Iout = Iout
+        self.Iout_count = Iout_count
+        self.Iout_track = Iout_track
```
[the remark this answers is not in the document]
Comment (KatieBuc) [reply sktime/annotation/hidalgo.py:238]: should I change from `N` to `_N` everywhere too?
Or just 
`self._N = N`

and then when using this value elsewhere
`N = self._N`

?

File: sktime/annotation/hidalgo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aiwalter) [reply sktime/annotation/hidalgo.py:17]: I meant having the class before the methods and functions in the file

File: sktime/annotation/hidalgo.py (hunk 1 above)
Comment (aiwalter) [reply sktime/annotation/hidalgo.py]: into `predict()`, or `transform()` if we go for a transformer

File: sktime/transformations/series/hidalgo.py [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+from sklearn.utils.validation import check_random_state
+
+from sktime.transformations.base import BaseTransformer
+
+
+class Hidalgo(BaseTransformer):
+    """Class of the Hidalgo intrinsic dimension model.
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:19]:
```suggestion
    """Heteregeneous Intrinsic Dimensionality Algorithm (Hidalgo) model.
```

File: sktime/transformations/series/hidalgo.py [hunk 8]
```diff
@@ -0,0 +1,672 @@
...
+        z = (z_1, ..., z_K) is a latent variable introduced, where z_i = k
+        indicates point i belongs to manifold K
+    use_Potts : bool, optional, default=True
+        if using local interaction between z, see equation (4)
+    estimate_zeta : bool, optional, default=False
+        update zeta in the sampling
+    sampling_rate: int, optional, default=10
+        rate at which to save samples for each n_iter
+    a : np.ArrayLike, optional, default=None
+        prior parameters of d
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:56]: shouldn't it be

```suggestion
        prior parameters of a
```

File: sktime/transformations/series/hidalgo.py [hunk 9]
```diff
@@ -0,0 +1,672 @@
...
+                        frac = np.exp(-b1[k] * r1)
+
+                    if frac > r2:
+                        stop = True
+                        d[k] = r1
+
+            return d
+
+        def sample_p(K, p, pp, c1):
+
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:360]: missing docstring

File: sktime/transformations/series/hidalgo.py [hunk 10]
```diff
@@ -0,0 +1,672 @@
...
+        fixed_Z = self.fixed_Z
+        use_Potts = self.use_Potts
+        estimate_zeta = self.estimate_zeta
+
+        sampling = np.empty(shape=0)
+        pp = (K - 1) / K
+        p = np.ones(shape=K) / K
+
+        def sample_d(K, a1, b1):
+
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:333]: missing docstring

File: sktime/transformations/series/hidalgo.py [hunk 11]
```diff
@@ -0,0 +1,672 @@
...
+                        stop = True
+                        r1 = r1 * (1.0 - pp + p[k])
+                        p[K - 1] += p[k] - r1
+                        pp -= p[k] - r1
+                        p[k] = r1
+
+            return (p, pp)
+
+        def sample_zeta(K, zeta, use_Potts, estimate_zeta, q, NN, f1, it):
+
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:383]: missing docstring

File: sktime/transformations/series/hidalgo.py [hunk 12]
```diff
@@ -0,0 +1,672 @@
...
+        self.Pi = Pi / np.shape(bestsampling)[0]
+        Z = np.argmax(Pi, axis=0)
+        pZ = np.max(Pi, axis=0)
+        Z[np.where(pZ < 0.8)] = -1
+
+        return Z
+
+
+def binom(N, q):
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:633]:
```suggestion
def binom(N: Union[int, float], q: Union[int, float]):
```

File: sktime/transformations/series/hidalgo.py [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+        number of subset q elements chosen from N
+    """
+    if q == 0:
+        return 1.0
+    if N < 0:
+        return 0.0
+    return reduce(lambda x, y: x * y, [(N - q1) / (q1 + 1) for q1 in range(q)])
+
+
+def Zpart(N, N1, zeta, q):
```
Comment (lmmentel) [inline sktime/transformations/series/hidalgo.py:650]: consider renaming the function to `partition_function`

**Comment by KatieBuc** [comment]
Tests are passing! Okay, so this is Hidalgo as a Transformer.

A couple of questions:

1. Since `predict` can't write to self, how else can we get important algorithmic information? There are various outputs from Gibbs sampling, i.e.
```
        # _d = np.mean(bestsampling[:, :K], axis=0)
        # _derr = np.std(bestsampling[:, :K], axis=0)
        # _p = np.mean(bestsampling[:, K : 2 * K], axis=0)
        # _perr = np.std(bestsampling[:, K : 2 * K], axis=0)
        # _lik = np.mean(bestsampling[:, -1], axis=0)
        # _likerr = np.std(bestsampling[:, -1], axis=0)
```
which are commented out for now - any ideas?

2. depending on the operating system, the dtype of an array was either printed, or not - which meant the doc string example matched depending on the operating system i.e.
```
PASSED: Install and test / test-windows (3.9) (pull_request) 

FAILED: Install and test / test-unix (3.8, ubuntu-20.04) (pull_request) 

2022-08-23T10:40:15.1569321Z ___________ [doctest] sktime.transformations.series.hidalgo.Hidalgo ____________
2022-08-23T10:40:15.1570690Z [gw1] linux -- Python 3.8.13 /opt/hostedtoolcache/Python/3.8.13/x64/bin/python
2022-08-23T10:40:15.1570983Z 074     Examples
2022-08-23T10:40:15.1571236Z 075     --------
2022-08-23T10:40:15.1571528Z 076     >>> from sktime.transformations.series.hidalgo import Hidalgo
2022-08-23T10:40:15.1571842Z 077     >>> import numpy as np
2022-08-23T10:40:15.1572080Z 078     >>> np.random.seed(123)
2022-08-23T10:40:15.1572319Z 079     >>> X = np.random.rand(10,3)
2022-08-23T10:40:15.1572548Z 080     >>> X[6:, 1:] = 0
2022-08-23T10:40:15.1572810Z 081     >>> model = Hidalgo(K=2, burn_in=0.5, n_iter=50, seed=10)
2022-08-23T10:40:15.1573081Z 082     >>> Z = model.fit_transform(X)
2022-08-23T10:40:15.1573288Z 083     >>> Z
2022-08-23T10:40:15.1573483Z Expected:
2022-08-23T10:40:15.1573885Z     array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1], dtype=int64)
2022-08-23T10:40:15.1574106Z Got:
2022-08-23T10:40:15.1574297Z     array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1])
```
so I made it `Z.tolist()`, hoping that's okay? open to other suggestions!

@fkiraly @lmmentel @aiwalter @Lovkush-A @miraep8

**Comment by KatieBuc** [comment]
also @lmmentel I think it's pretty trivial to make `gibbs_sampling` a non-class method, if we just pass in
```
        zeta = self.zeta
        q = self.q
        K = self.K
        n_iter = self.n_iter
        fixed_Z = self.fixed_Z
        use_Potts = self.use_Potts
        estimate_zeta = self.estimate_zeta
```
these variables instead, and a couple of those then into `_update_zeta_prior`, could do this if you think it's important?

**Comment by aiwalter** [comment]
nice @KatieBuc 🎉 

> Since predict can't write to self, how else can we get important algorithmic information?

I guess you mean `transform`? Good question if `fit` is empty, maybe we need to make an exception here that we allow state changes, what do others think?

> depending on the operating system, the dtype of an array was either printed, or not - which meant the doc string example matched depending on the operating system i.e.

Is it possible that you use one of our sample data, e.g. `load_airline`, `load_shampoo_sales`, `load_longley`? Then the output would be `pandas`. If you want to show output, it might help to shorten the data or just take `load_longley`, its quite short so output will be short as well.

File: sktime/transformations/series/hidalgo.py [hunk 14]
```diff
@@ -0,0 +1,736 @@
...
+
+        a1 = a + NN
+        b1 = b + V
+        c1 = c + NN
+
+        N_in, f1 = self._update_zeta_prior(Z, N, Iin)
+
+        return (V, NN, a1, b1, c1, Z, f1, N_in)
+
+    def gibbs_sampling(
```
Comment (aiwalter) [inline sktime/transformations/series/hidalgo.py:261]:
```suggestion
    def _gibbs_sampling(
```

**Review by aiwalter — CHANGES_REQUESTED** [review]

**Comment by fkiraly** [comment]
> Since `_transform` can't write to self, how else can we get important algorithmic information?

A simple solution to this problem I've used before:
* run the parts that write to `self` in `_fit`
* in `_transform`, check whether the data is identical to the data seen in `fit`. If yes, run the rest. If not, run the entire algorithm, but do *not* write to `self`.

**Comment by KatieBuc** [comment]
> A simple solution to this problem I've used before:
> 
> * run the parts that write to `self` in `_fit`
> * in `_transform`, check whether the data is identical to the data seen in `fit`. If yes, run the rest. If not, run the entire algorithm, but do _not_ write to `self`.

Hmm in this case, if we separate logic into fit and predict/transform, then the data _must_ be identical (and errors will/should be thrown otherwise). Also, the commented out variables are the last task of `_transform`, there's really nothing else to run anyway - it would be cool to have a kind of `get_params`/`set_params` but for auxiliary outputs; parameters that are found in predict or transform that are of interest for subsequent analysis. What do you think?

What is the simplest solution to this for now? Even if we put all the logic in `_fit`, and refit to ensure the algorithm can run, we still don't get the fitted_model returned i.e.

```
def _transform(...):
    if np.allclose(X,fitted_model.X):
        return fitted_model.Z
    fitted_model = _fit(...)
    return fitted_model.Z
```
@fkiraly

**Comment by KatieBuc** [comment]
@fkiraly ignore my last comment^^

just put everything in fit, see last commit - should be fine now? (i.e. no separation of logic, passing required unit tests, predict is basically just selecting what to return from the fitted object)

**Review by fkiraly — APPROVED** [review]
Merging as current, looks good - well done!

We discussed potential additional changes, to be made in separate PR:
* if data in `transform` is different from `fit`, re-apply `fit` of a clone, instead of just passing the output of `fit`
* allow arbitrary `sktime` distance to be passed and used internally

**Comment by lmmentel** [comment]
:tada:
